#TODO: 1 tests
import os
import unittest

from convert_gvf_to_vcf.assistingconverter import get_gvf_attributes, generate_custom_structured_meta_line, \
    convert_gvf_attributes_to_vcf_values
from convert_gvf_to_vcf.utils import read_yaml, generate_symbolic_allele_dict
from convert_gvf_to_vcf.lookup import Lookup

class TestAssistingConverter(unittest.TestCase):
    def setUp(self):
        input_folder = os.path.dirname(__file__)
        self.input_file = os.path.join(input_folder, "input", "zebrafish.gvf")
        self.input_folder_parent = os.path.abspath(os.path.join(os.path.dirname( __file__ ), '..', 'convert_gvf_to_vcf'))
        # the inputs below are INFO attribute files
        self.etc_folder =  os.path.join(self.input_folder_parent, "etc")
        self.assembly = os.path.join(input_folder, "input", "zebrafish.fa")
        self.etc_folder =  os.path.join(self.input_folder_parent, "etc")
        self.output_file = os.path.join(input_folder, "input", "a.vcf")
        self.reference_lookup = Lookup(self.assembly)

    def test_generate_custom_structured_meta_line(self):
        field = "INFO"
        idkey = "END"
        number = "1"
        data_type = "Integer"
        description = "Testing the end position"
        # Testing None
        optional_data_none = None
        output_for_none = generate_custom_structured_meta_line(
            field,
            idkey,
            number,
            data_type,
            description,
            optional_data_none
        )
        expected_output_for_none = "##INFO=<ID=END,Number=1,Type=Integer,Description=\"Testing the end position\">"
        assert output_for_none == expected_output_for_none
        # Testing optional data
        optional_data = {"test1": "A"}
        output_for_optional = generate_custom_structured_meta_line(
            field,
            idkey,
            number,
            data_type,
            description,
            optional_data
        )
        expected_output_for_optional = "##INFO=<ID=END,Number=1,Type=Integer,Description=\"Testing the end position\",test1=\"A\">"
        assert output_for_optional == expected_output_for_optional

    def test_get_gvf_attributes(self):
        col9 = "ID=1;Name=nssv1412199;Alias=CNV28955;parent=nsv811094;Start_range=.,1;End_range=2,.;sample_name=Wilds2-3"
        gvf_attribute_dictionary = get_gvf_attributes(col9)
        expected_dictionary = {'ID': '1', 'Name': 'nssv1412199', 'Alias': 'CNV28955', 'parent': 'nsv811094', 'Start_range': ['.', '1'], 'End_range': ['2', '.'], 'sample_name': 'Wilds2-3'}
        assert gvf_attribute_dictionary == expected_dictionary

    def test_convert_gvf_attributes_to_vcf_values(self):
        col9 = "ID=1;Name=nssv1412199;Alias=CNV28955;parent=nsv811094;Start_range=.,1;End_range=2,.;sample_name=Wilds2-3;Genotype=0:1"
        # self.reference_lookup.mapping_attribute_dict
        field_lines_dictionary = {'ALT': [], 'INFO': [], 'FILTER': [], 'FORMAT': []}
        all_possible_lines_dictionary = {'ALT': {'DEL:ME:ALU': '##ALT=<ID=DEL:ME:ALU,Description="Alu deletion">', 'INS': '##ALT=<ID=INS,Description="Novel sequence insertion">', 'None': '##ALT=<ID=None,Description="Sequence alteration">', 'DUP': '##ALT=<ID=DUP,Description="Duplication">', 'DEL': '##ALT=<ID=DEL,Description="Deletion">', 'CNV': '##ALT=<ID=CNV,Description="Copy number variable region">', 'BND': '##ALT=<ID=BND,Description="Interchromosomal translocation">', 'INV': '##ALT=<ID=INV,Description="Inversion">', 'DEL:ME:LINE1': '##ALT=<ID=DEL:ME:LINE1,Description="LINE1 deletion">', 'INS:ME:LINE1': '##ALT=<ID=INS:ME:LINE1,Description="LINE1 Insertion">', 'INS:ME': '##ALT=<ID=INS:ME,Description="Mobile element insertion">', 'CNV:TR': '##ALT=<ID=CNV:TR,Description="Tandem repeat determined based on DNA abundance">', 'INS:ME:SVA': '##ALT=<ID=INS:ME:SVA,Description="SVA insertion">', 'DUP:TANDEM': '##ALT=<ID=DUP:TANDEM,Description="Tandem duplication">'}, 'INFO': {'AA': '##INFO=<ID=AA,Number=1,Type=String,Description="Ancestral allele">', 'AC': '##INFO=<ID=AC,Number=A,Type=Integer,Description="Allele count in genotypes, for each ALT allele, in the same order as listed">', 'ADF': '##INFO=<ID=ADF,Number=R,Type=Integer,Description="Read depth for each allele on the forward strand">', 'ADR': '##INFO=<ID=ADR,Number=R,Type=Integer,Description="Read depth for each allele on the reverse strand">', 'AF': '##INFO=<ID=AF,Number=A,Type=Float,Description="Allele frequency for each ALT allele in the same order as listed (estimated from primary data, not called genotypes)">', 'AN': '##INFO=<ID=AN,Number=1,Type=Integer,Description="Total number of alleles in called genotypes">', 'BQ': '##INFO=<ID=BQ,Number=1,Type=Float,Description="RMS base quality">', 'BKPTID': '##INFO=<ID=BKPTID,Number=A,Type=String,Description="ID of the assembled alternate allele in the assembly file">', 'CICN': '##INFO=<ID=CICN,Number=.,Type=Float,Description="Confidence interval around copy number">', 'CIEND': '##INFO=<ID=CIEND,Number=.,Type=Integer,Description="Confidence interval around END for symbolic structural variants">', 'CILEN': '##INFO=<ID=CILEN,Number=.,Type=Integer,Description="Confidence interval for the SVLEN field">', 'CIRB': '##INFO=<ID=CIRB,Number=.,Type=Integer,Description="Confidence interval around RB">', 'CIRUC': '##INFO=<ID=CIRUC,Number=.,Type=Float,Description="Confidence interval around RUC">', 'CIPOS': '##INFO=<ID=CIPOS,Number=.,Type=Integer,Description="Confidence interval around POS for symbolic structural variants">', 'CIGAR': '##INFO=<ID=CIGAR,Number=A,Type=String,Description="Cigar string describing how to align an alternate allele to the reference allele">', 'CN': '##INFO=<ID=CN,Number=A,Type=Float,Description="Copy number of CNV/breakpoint">', 'DB': '##INFO=<ID=DB,Number=0,Type=Flag,Description="dbSNP membership">', 'DBRIPID': '##INFO=<ID=DBRIPID,Number=A,Type=String,Description="ID of this element in DBRIP">', 'DBVARID': '##INFO=<ID=DBVARID,Number=A,Type=String,Description="ID of this element in DBVAR">', 'DGVID': '##INFO=<ID=DGVID,Number=A,Type=String,Description="ID of this element in Database of Genomic Variation">', 'DP': '##INFO=<ID=DP,Number=1,Type=Integer,Description="Combined depth across samples">', 'END': '##INFO=<ID=END,Number=1,Type=Integer,Description="End position on CHROM (used with symbolic alleles; see below) or End position of the longest variant described in this record">', 'EVENT': '##INFO=<ID=EVENT,Number=A,Type=String,Description="ID of associated event">', 'EVENTTYPE': '##INFO=<ID=EVENTTYPE,Number=A,Type=String,Description="Type of associated event">', 'H2': '##INFO=<ID=H2,Number=0,Type=Flag,Description="HapMap2 membership">', 'H3': '##INFO=<ID=H3,Number=0,Type=Flag,Description="HapMap3 membership">', 'HOMLEN': '##INFO=<ID=HOMLEN,Number=A,Type=Integer,Description="Length of base pair identical micro-homology at breakpoints">', 'HOMSEQ': '##INFO=<ID=HOMSEQ,Number=A,Type=String,Description="Sequence of base pair identical micro-homology at breakpoints">', 'IMPRECISE': '##INFO=<ID=IMPRECISE,Number=0,Type=Flag,Description="Imprecise structural variation">', 'MQ': '##INFO=<ID=MQ,Number=1,Type=Float,Description="RMS mapping quality">', 'MQ0': '##INFO=<ID=MQ0,Number=1,Type=Integer,Description="Number of MAPQ == 0 reads">', 'MATEID': '##INFO=<ID=MATEID,Number=A,Type=String,Description="ID of mate breakend">', 'MEINFO': '##INFO=<ID=MEINFO,Number=.,Type=String,Description="Mobile element info of the form NAME,START,END,POLARITY">', 'METRANS': '##INFO=<ID=METRANS,Number=.,Type=String,Description="Mobile element transduction info of the form CHR,START,END,POLARITY">', 'NOVEL': '##INFO=<ID=NOVEL,Number=0,Type=Flag,Description="Indicates a novel structural variation">', 'NS': '##INFO=<ID=NS,Number=1,Type=Integer,Description="Number of samples with data">', 'PARID': '##INFO=<ID=PARID,Number=A,Type=String,Description="ID of partner breakend">', 'RN': '##INFO=<ID=RN,Number=A,Type=Integer,Description="Total number of repeat sequences in this allele">', 'RUS': '##INFO=<ID=RUS,Number=.,Type=String,Description="Repeat unit sequence of the corresponding repeat sequence">', 'RUL': '##INFO=<ID=RUL,Number=.,Type=Integer,Description="Repeat unit length of the corresponding repeat sequence">', 'RUC': '##INFO=<ID=RUC,Number=.,Type=Float,Description="Repeat unit count of corresponding repeat sequence">', 'RB': '##INFO=<ID=RB,Number=.,Type=Integer,Description="Total number of bases in the corresponding repeat sequence">', 'RUB': '##INFO=<ID=RUB,Number=.,Type=Integer,Description="Number of bases in each individual repeat unit">', 'SB': '##INFO=<ID=SB,Number=4,Type=Integer,Description="Strand bias">', 'SOMATIC': '##INFO=<ID=SOMATIC,Number=0,Type=Flag,Description="Somatic mutation (for cancer genomics)">', 'SVCLAIM': '##INFO=<ID=SVCLAIM,Number=A,Type=String,Description="Claim made by the structural variant call. Valid values are D, J, DJ for abundance, adjacency and both respectively">', 'SVLEN': '##INFO=<ID=SVLEN,Number=A,Type=String,Description="Length of structural variant">', 'SVTYPE': '##INFO=<ID=SVTYPE,Number=1,Type=String,Description="Type of structural variant">', 'VALIDATED': '##INFO=<ID=VALIDATED,Number=0,Type=Flag,Description="Validated by follow-up experiment">', '1000G': '##INFO=<ID=1000G,Number=0,Type=Flag,Description="1000 Genomes membership">', 'AD': '##INFO=<ID=AD,Number=R,Type=Integer,Description="Total read depth for each allele">', 'TIFL': '##INFO=<ID=TIFL,Number=.,Type=String,Description="Three prime inner flank link">', 'FIFL': '##INFO=<ID=FIFL,Number=.,Type=String,Description="Five prime inner flank link">', 'AM': '##INFO=<ID=AM,Number=.,Type=String,Description="Assertion method">', 'BO': '##INFO=<ID=BO,Number=.,Type=String,Description="breakpoint order">', 'CLINSIG': '##INFO=<ID=CLINSIG,Number=.,Type=String,Description="Clinical Significance">', 'ESL': '##INFO=<ID=ESL,Number=.,Type=String,Description="Evidence Sequence Link">', 'LOG': '##INFO=<ID=LOG,Number=.,Type=Float,Description="Log2 Value">', 'MO': '##INFO=<ID=MO,Number=.,Type=Integer,Description="Mutation">', 'NAME': '##INFO=<ID=NAME,Number=.,Type=String,Description="Name">', 'PARENT': '##INFO=<ID=PARENT,Number=.,Type=String,Description="Parent">', 'PHENODESC': '##INFO=<ID=PHENODESC,Number=.,Type=Stromg,Description="Phenotype description">', 'PHENOLINK': '##INFO=<ID=PHENOLINK,Number=.,Type=String,Description="Phenotype link">', 'RALIGN': '##INFO=<ID=RALIGN,Number=.,Type=String,Description="Reciprocal Alignment">', 'REMAP': '##INFO=<ID=REMAP,Number=.,Type=Float,Description="Remap score">', 'SAMP': '##INFO=<ID=SAMP,Number=.,Type=String,Description="Samples">', 'SVCID': '##INFO=<ID=SVCID,Number=.,Type=Integer,Description="submitter variant call ID">', 'SVRID': '##INFO=<ID=SVRID,Number=.,Type=Integer,Description="Submitter variant region ID">', 'SUPCOUNT': '##INFO=<ID=SUPCOUNT,Number=.,Type=Integer,Description="Support Count">', 'SSEQLINK': '##INFO=<ID=SSEQLINK,Number=.,Type=String,Description="Supporting Sequence Link">', 'VARCALLDESC': '##INFO=<ID=VARCALLDESC,Number=.,Type=String,Description="Variant Call Description">', 'VARREGSOID': '##INFO=<ID=VARREGSOID,Number=.,Type=String,Description="Variant Region Sequence Ontology ID">', 'VARCALLSOID': '##INFO=<ID=VARCALLSOID,Number=.,Type=String,Description="Variant call Sequence ontology ID">', 'ID': '##INFO=<ID=ID,Number=.,Type=String,Description="A unique identifier">', 'VARSEQ': '##INFO=<ID=VARSEQ,Number=.,Type=String,Description="Alleles found in an individual (or group of individuals).">', 'REFSEQ': '##INFO=<ID=REFSEQ,Number=.,Type=String,Description="The sequence corresponding to the seq id, start, end of the associated genomic sequence file.">', 'ALIAS': '##INFO=<ID=ALIAS,Number=.,Type=String,Description="Secondary Name">', 'VAREFF': '##INFO=<ID=VAREFF,Number=.,Type=String,Description="The effect of a sequence alteration on a sequence feature">', 'VARCDN': '##INFO=<ID=VARCDN,Number=.,Type=String,Description="The codons that overlap a feature as modified by the sequences in the Variant_seq attributes">', 'REFCDN': '##INFO=<ID=REFCDN,Number=.,Type=String,Description="The codons that overlap a feature on the the reference genome.">', 'VARAA': '##INFO=<ID=VARAA,Number=.,Type=String,Description="The amino acids that overlap a feature as modified by the sequences in the Variant_seq attribute.">', 'REFAA': '##INFO=<ID=REFAA,Number=.,Type=String,Description="The amino acids that overlap a feature on the reference genome.">', 'BD': '##INFO=<ID=BD,Number=.,Type=String,Description="Describes the source or destination of the a zero-length sequence alteration in the form seqid:start-end:strand">', 'SEQCTXT': '##INFO=<ID=SEQCTXT,Number=.,Type=String,Description="A region of sequence context from the reference genome around a sequence alteration.">', 'DBXREF': '##INFO=<ID=DBXREF,Number=.,Type=String,Description="A database cross-reference">'}, 'FILTER': {}, 'FORMAT': {'AHAP': '##FORMAT=<ID=AHAP,Number=1,Type=Integer,Description="Unique identifier of ancestral haplotype">', 'ADF': '##FORMAT=<ID=ADF,Number=R,Type=Integer,Description="Read depth for each allele on the forward strand">', 'ADR': '##FORMAT=<ID=ADR,Number=R,Type=Integer,Description="Read depth for each allele on the reverse strand">', 'CICN': '##FORMAT=<ID=CICN,Number=2,Type=Float,Description="Confidence interval around copy number">', 'CN': '##FORMAT=<ID=CN,Number=1,Type=Float,Description="CopyNumber">', 'CNL': '##FORMAT=<ID=CNL,Number=G,Type=Float,Description="Copy number genotype likelihood">', 'CNP': '##FORMAT=<ID=CNP,Number=G,Type=Float,Description="Copy number posterior probabilities">', 'CNQ': '##FORMAT=<ID=CNQ,Number=1,Type=Float,Description="Copy number genotype quality">', 'DP': '##FORMAT=<ID=DP,Number=1,Type=Integer,Description="Read depth">', 'EC': '##FORMAT=<ID=EC,Number=A,Type=Integer,Description="Expected alternate allele counts">', 'FT': '##FORMAT=<ID=FT,Number=1,Type=String,Description="Filter indicating if this genotype was \'called\'">', 'GL': '##FORMAT=<ID=GL,Number=G,Type=Float,Description="Genotype likelihoods">', 'GP': '##FORMAT=<ID=GP,Number=G,Type=Float,Description="Genotype posterior probabilities">', 'GQ': '##FORMAT=<ID=GQ,Number=1,Type=Integer,Description="Conditional genotype quality">', 'HAP': '##FORMAT=<ID=HAP,Number=1,Type=Integer,Description="Unique haplotype identifier">', 'HQ': '##FORMAT=<ID=HQ,Number=2,Type=Integer,Description="Haplotype quality">', 'MQ': '##FORMAT=<ID=MQ,Number=1,Type=Integer,Description="RMS mapping quality">', 'NQ': '##FORMAT=<ID=NQ,Number=1,Type=Integer,Description="Phred style probability score that the variant is novel">', 'PL': '##FORMAT=<ID=PL,Number=G,Type=Integer,Description="Phred-scaled genotype likelihoods rounded to the closest integer">', 'PP': '##FORMAT=<ID=PP,Number=G,Type=Integer,Description="Phred-scaled genotype posterior probabilities rounded to the closest integer">', 'PQ': '##FORMAT=<ID=PQ,Number=1,Type=Integer,Description="Phasing quality">', 'GT': '##FORMAT=<ID=GT,Number=1,Type=String,Description="Zygosity">', 'PS': '##FORMAT=<ID=PS,Number=1,Type=Integer,Description="Phase set">', 'AD': '##FORMAT=<ID=AD,Number=R,Type=Integer,Description="Read depth for each allele">'}}
        gvf_attribute_dictionary, vcf_info_values, vcf_format_values = convert_gvf_attributes_to_vcf_values(
            col9,
            self.reference_lookup.mapping_attribute_dict,
            field_lines_dictionary,
            all_possible_lines_dictionary
        )
        # testing gvf_attribute_dictionary
        expected_gvf_attribute_dictionary = {'ID': '1', 'Name': 'nssv1412199', 'Alias': 'CNV28955', 'parent': 'nsv811094', 'Start_range': ['.', '1'], 'End_range': ['2', '.'], 'sample_name': 'Wilds2-3', 'Genotype': '0:1'}
        assert gvf_attribute_dictionary == expected_gvf_attribute_dictionary
        # testing vcf_info_values
        expected_info_values = {'ID': '1', 'NAME': 'nssv1412199', 'ALIAS': 'CNV28955'}
        assert vcf_info_values == expected_info_values
        # testing vcf_format_values
        expected_format_values = {'Wilds2-3': {'GT': '0:1'}}
        assert vcf_format_values == expected_format_values